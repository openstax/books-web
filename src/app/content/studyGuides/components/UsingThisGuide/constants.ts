export const cookieUTG = 'user_has_seen_UTG';
