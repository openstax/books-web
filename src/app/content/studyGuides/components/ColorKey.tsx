import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import colorKeyIcon from '../../../../assets/colorKeyIcon.png';
import { useAnalyticsEvent } from '../../../../helpers/analytics';
import { PlainButton } from '../../../components/Button';
import { textRegularStyle } from '../../../components/Typography';
import theme from '../../../theme';
import { toolbarIconStyles } from '../../components/Toolbar/iconStyles';
import { toolbarDefaultText } from '../../components/Toolbar/styled';
import ColorIndicator from '../../highlights/components/ColorIndicator';
import { mobilePaddingSides } from '../../styles/PopupConstants';
import { highlightStyles } from '../constants';

const noteConstants = {
  noteHeight: 16.8,
  noteWidth: 36,
};

// tslint:disable-next-line:variable-name
export const ColorKeyButtonWrapper = styled(PlainButton)`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: auto;
  padding-right: 2.4rem;
  position: relative;
  overflow: visible;
`;

// tslint:disable-next-line:variable-name
const ColorKeyDescription = styled.div`
  ::before {
    content: "";
    position: absolute;
    width: 1.6rem;
    height: 1.6rem;
    transform: rotate(45deg);
    top: -0.9rem;
    right: 5.5rem;
    z-index: 1;
    background: ${theme.color.white};
    border-top: 0.1rem solid ${theme.color.neutral.formBorder};
    border-left: 0.1rem solid ${theme.color.neutral.formBorder};

    ${theme.breakpoints.mobile(css`
      right: 2rem;
    `)}
  }

  min-height: ${noteConstants.noteHeight}rem;
  width: ${noteConstants.noteWidth}rem;
  box-sizing: border-box;
  border: 0.1rem solid ${theme.color.neutral.formBorder};
  box-shadow: 0 0.4rem 0.6rem 0 rgba(0,0,0,0.2);
  position: absolute;
  top: 4rem;
  right: 5rem;
  background: ${theme.color.white};
  overflow: visible;
  padding: 1.5rem 1.65rem;

  ${theme.breakpoints.mobile(css`
    width: calc(100vw - ${mobilePaddingSides * 2}rem);
    right: ${mobilePaddingSides / 2}rem;
  `)}
`;

// tslint:disable-next-line:variable-name
const ColorKeyIcon = styled.img`
  ${toolbarIconStyles}
`;

// tslint:disable-next-line:variable-name
const ColorKeyText = styled.span`
  ${toolbarDefaultText}
`;

// tslint:disable-next-line:variable-name
const KeyTermText = styled.span`
  ${textRegularStyle}
  font-size: 1.4rem;
  letter-spacing: -0.4px;
  margin-left: 1.6rem;
  width: 100%;
  text-align: left;
`;

// tslint:disable-next-line:variable-name
const KeyTermWrapper = styled.div`
  display: flex;
  align-items: flex-start;
  padding: 0.5rem 0;

  ${ColorIndicator} {
    margin-top: 0.5rem;
    width: 1.8rem;
  }
`;

// tslint:disable-next-line:variable-name
const ColorKey = () => {
  const [open, setOpen] = React.useState<boolean>(false);
  const trackOpenClose = useAnalyticsEvent('openCloseColorKey');

  const toggleColorKey = () => {
    setOpen((state) => !state);
    trackOpenClose(open);
  };

  return <FormattedMessage id='i18n:studyguides:popup:color-key'>
    {(msg: Element | string) =>
      <ColorKeyButtonWrapper onClick={toggleColorKey} aria-label={msg}>
        <ColorKeyIcon src={colorKeyIcon}/>
        <ColorKeyText>{msg}</ColorKeyText>
        {open && <ColorKeyDescription>
          {highlightStyles.map((color, index) =>
            <KeyTermWrapper key={index}>
              <FormattedMessage id='i18n:studyguides:popup:color-key:terms:aria-label' values={{color: color.label}}>
                {(arialabel: string) => <ColorIndicator
                  shape='circle'
                  style={color}
                  size='small'
                  aria-label={arialabel}
                  component={<label />}
                />}
              </FormattedMessage>
              <FormattedMessage id={'i18n:studyguides:popup:color-key:terms:' + color.label}>
                {(key: string) => <KeyTermText>{key}</KeyTermText>}
              </FormattedMessage>
            </KeyTermWrapper>
          )
        }
        </ColorKeyDescription>}
      </ColorKeyButtonWrapper>
    }
  </FormattedMessage>;
};

export default ColorKey;
