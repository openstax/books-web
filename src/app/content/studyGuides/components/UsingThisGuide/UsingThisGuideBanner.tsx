import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import { Times } from 'styled-icons/fa-solid/Times';
import { PlainButton } from '../../../../components/Button';
import { H2, h4MobileStyle } from '../../../../components/Typography/headings';
import theme from '../../../../theme';
import { filters } from '../../../styles/PopupConstants';
import desktopBanner from './assets/banner.png';
import mobileBanner from './assets/banner_mobile.png';

// tslint:disable-next-line: variable-name
const BannerWrapper = styled.div`
  outline: none;
  position: relative;
  padding: ${filters.dropdownToggle.topBottom.desktop}rem ${filters.dropdownToggle.sides.desktop}rem;
  width: 100%;
  background: ${theme.color.black};
  margin-bottom: 1rem;
  ${theme.breakpoints.mobileSmall(css`
    padding: ${filters.dropdownToggle.topBottom.mobile}rem ${filters.dropdownToggle.sides.mobile}rem;
  `)}
`;

// tslint:disable-next-line: variable-name
export const DesktopBanner = styled.img`
  width: 100%;
  padding: 0 4.2rem;
  ${theme.breakpoints.mobileMedium(css`
    max-width: 30rem;
    padding: 0;
  `)}
`;

// tslint:disable-next-line: variable-name
const UsingThisGuideTitle = styled(H2)`
  text-align: center;
  width: 100%;
  color: ${theme.color.white};
  ${theme.breakpoints.mobile(css`
    ${h4MobileStyle}
    color: ${theme.color.white};
  `)}
`;

// tslint:disable-next-line:variable-name
export const CloseIconButton = styled(PlainButton)`
  position: absolute;
  border-radius: 50%;
  background: ${theme.color.white};
  display: flex;
  align-items: center;
  justify-content: center;
  top: ${filters.dropdownToggle.topBottom.desktop}rem;
  right: ${filters.dropdownToggle.sides.desktop}rem;
  ${theme.breakpoints.mobileSmall(css`
    top: ${filters.dropdownToggle.topBottom.mobile}rem;
    right: ${filters.dropdownToggle.sides.mobile}rem;
  `)}
`;

// tslint:disable-next-line:variable-name
export const CloseIcon = styled(Times)`
  color: ${theme.color.black};
  height: 2.8rem;
  width: 2.8rem;
  padding: 0.4rem;
  ${theme.breakpoints.mobile(css`
    height: 1.6rem;
    width: 1.6rem;
  `)}
`;

// tslint:disable-next-line: variable-name
const HeaderWrapper = styled.div`
  display: flex;
  align-items: center;
`;

// tslint:disable-next-line: variable-name
const BodyWrapper = styled.div`
  display: flex;
  justify-content: center;
`;

interface Props {
  onClick: () => void;
  isUTGopen: boolean;
}

// tslint:disable-next-line:variable-name
const UsingThisGuideBanner = (props: Props) => {
  return <BannerWrapper>
    <HeaderWrapper>
      <FormattedMessage id='i18n:studyguides:popup:using-this-guide'>
        {(msg: Element | string) => <UsingThisGuideTitle>{msg}</UsingThisGuideTitle>}
      </FormattedMessage>
    </HeaderWrapper>
    <BodyWrapper>
      <FormattedMessage id='i18n:studyguides:popup:using-this-guide:alt'>
        {(msg: Element | string) => <picture>
          <source media={theme.breakpoints.mobileMediumQuery} srcSet={mobileBanner} />
          <DesktopBanner src={desktopBanner} alt={msg} tabIndex={0}/>
        </picture>}
      </FormattedMessage>
    </BodyWrapper>
    <CloseIconButton onClick={props.onClick}>
      <CloseIcon/>
    </CloseIconButton>
  </BannerWrapper>;
};

export default UsingThisGuideBanner;
