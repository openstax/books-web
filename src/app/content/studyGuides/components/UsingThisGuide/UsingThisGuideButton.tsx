import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import { PlainButton } from '../../../../components/Button';
import theme from '../../../../theme';
import { toolbarDefaultText } from '../../../components/Toolbar/styled';
import { QuestionCircle } from 'styled-icons/fa-regular/QuestionCircle';
import { mobilePaddingSides } from '../../../styles/PopupConstants';

const UTGButton = {
  height: 6,
  marginBottom: 0.6,
  negativeMarginBottom: 0.1
}

// tslint:disable-next-line:variable-name
const QuestionIcon = styled(QuestionCircle)`
  height: 1.6rem;
  width: 1.6rem;
`;

// tslint:disable-next-line:variable-name
export const UsingThisGuideButtonWrapper = styled(PlainButton)`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: auto;
  position: relative;
  overflow: visible;
  color: ${(props) => props.isOpen ? theme.color.white : theme.color.primary.gray.base};
  background: ${(props) => props.isOpen ? 'black' : 'white'};
  padding:  ${(props) => props.isOpen ? '0 1.6rem' : '0'};
  height: ${UTGButton.height - UTGButton.marginBottom}rem;
`;

// tslint:disable-next-line:variable-name
const UsingThisGuideText = styled.span`
  ${toolbarDefaultText}; 
`;

// tslint:disable-next-line:variable-name
const UsingThisGuideWrapper = styled.div`
  margin-right: 4.8rem;
  margin-bottom: ${(props) => props.isOpen ? -(UTGButton.marginBottom + UTGButton.negativeMarginBottom) : '0'}rem;
  ${theme.breakpoints.mobile(css`
    margin-right: ${mobilePaddingSides}rem;
  `)}
`;

interface Props {
  onClick: () => void;
  open: boolean;
}

// tslint:disable-next-line:variable-name
const UsingThisGuideButton = (props: Props) => {
  return <UsingThisGuideWrapper isOpen={props.open}>
    <FormattedMessage id='i18n:studyguides:popup:using-this-guide'>
      {(msg: Element | string) =>
        <UsingThisGuideButtonWrapper onClick={props.onClick} aria-label={msg} isOpen={props.open}>
          <QuestionIcon/>
          <UsingThisGuideText>{msg}</UsingThisGuideText>
        </UsingThisGuideButtonWrapper>
      }
    </FormattedMessage>
  </UsingThisGuideWrapper>;
};

export default UsingThisGuideButton;
