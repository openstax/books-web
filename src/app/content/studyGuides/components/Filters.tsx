import flow from 'lodash/fp/flow';
import React from 'react';
import { connect, useSelector } from 'react-redux';
import { loggedOut } from '../../../auth/selectors';
import { AppState, Dispatch } from '../../../types';
import ChapterFilter from '../../components/popUp/ChapterFilter';
import Filters, { FilterDropdown } from '../../components/popUp/Filters';
import FiltersList from '../../components/popUp/FiltersList';
import PrintButton from '../../components/popUp/PrintButton';
import { printStudyGuides, setSummaryFilters } from '../actions';
import * as selectors from '../selectors';
import ColorKey from './ColorKey';
import UsingThisGuideButton from './UsingThisGuide/UsingThisGuideButton';
import UsingThisGuideBanner from './UsingThisGuide/UsingThisGuideBanner';

// tslint:disable-next-line:variable-name
const ConnectedChapterFilter = connect(
  (state: AppState) => ({
    locationFilters: selectors.studyGuidesLocationFilters(state),
    locationFiltersWithContent: selectors.studyGuidesLocationFiltersWithContent(state),
    selectedLocationFilters: selectors.summaryLocationFilters(state),
  }),
  (dispatch: Dispatch) => ({
    setFilters: flow(setSummaryFilters, dispatch),
  })
)(ChapterFilter);

// tslint:disable-next-line:variable-name
const ConnectedFilterList = connect(
  (state: AppState) => ({
    locationFilters: selectors.studyGuidesLocationFilters(state),
    selectedLocationFilters: selectors.summaryLocationFilters(state),
  }),
  (dispatch: Dispatch) => ({
    setFilters: flow(setSummaryFilters, dispatch),
  })
)(FiltersList);

// tslint:disable-next-line:variable-name
const ConnectedPrintButton = connect(
  (state: AppState) => ({
    disabled: selectors.summaryIsLoading(state),
    loading: selectors.summaryIsLoading(state),
    shouldFetchMore: selectors.hasMoreResults(state),
  }),
  (dispatch: Dispatch) => ({
    loadHighlightsAndPrint: flow(printStudyGuides, dispatch),
  }),
  (stateProps, dispatchProps, ownProps) => {
    const {shouldFetchMore, loadHighlightsAndPrint, ...props} = {
      ...stateProps,
      ...dispatchProps,
      ...ownProps,
    };

    return shouldFetchMore
      ? {...props, onClick: loadHighlightsAndPrint}
      : props
    ;
  }
)(PrintButton);

export default () => {
  const userLoggedOut = useSelector(loggedOut);
  const [UTGopen, setUTGopen] = React.useState(false);

  const toggleUsingThisGuide = () => {
    setUTGopen((state) => !state);
  };

  const closeUsingThisGuide = React.useCallback(() => { setUTGopen(false); }, []);

  return <React.Fragment>
    <Filters>
      <FilterDropdown
        label='i18n:highlighting:filters:chapters'
        ariaLabelId='i18n:studyguides:popup:filters:filter-by:aria-label'
      >
        <ConnectedChapterFilter disabled={userLoggedOut}/>
      </FilterDropdown>
      <ColorKey />
      <UsingThisGuideButton onClick={toggleUsingThisGuide} open={UTGopen}/>
      <ConnectedPrintButton studyGuidesButton />
      {!userLoggedOut && <ConnectedFilterList />}
    </Filters>
    {UTGopen && <UsingThisGuideBanner onClick={closeUsingThisGuide}/>}
  </React.Fragment>;
};
