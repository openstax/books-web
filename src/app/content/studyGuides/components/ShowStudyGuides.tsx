import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import styled, { css } from 'styled-components/macro';
import GoToTopButton from '../../../components/GoToTopButton';
import theme from '../../../theme';
import { assertNotNull } from '../../../utils';
import FiltersList from '../../components/popUp/FiltersList';
import { loadMoreDistanceFromBottom } from '../../constants';
import { PopupBody } from '../../styles/PopupStyles';
import { loadMoreStudyGuides } from '../actions';
import * as select from '../selectors';
import Filters from './Filters';
import StudyGuides from './StudyGuides';
import StudyGuidesCTA from './StudyGuidesCTA';

// tslint:disable-next-line:variable-name
export const StudyGuidesBody = styled(PopupBody)`
  background: ${theme.color.neutral.darker};
  ${theme.breakpoints.mobile(css`
    text-align: left;
    padding: 0;
  `)}

  @media print {
    background: white;

    ${FiltersList} {
      padding-left: 0;
    }
  }
`;

// tslint:disable-next-line: variable-name
const ShowStudyGuides = () => {
  const ref = React.useRef<HTMLElement>(null);
  const [showGoToTop, setShowGoToTop] = React.useState(false);
  const dispatch = useDispatch();
  const isLoading = useSelector(select.summaryIsLoading);
  const hasMoreResults = useSelector(select.hasMoreResults);

  const goToTop = () => {
    assertNotNull(ref.current, 'Expected ref to be not null').scrollTop = 0;
    setShowGoToTop(false);
  };

  const updateBackToTop = (refElement: HTMLElement) => {
    setShowGoToTop(refElement.scrollTop > 0);
  };

  const fetchMoreStudyGuides = (refElement: HTMLElement) => {
    if (isLoading || !hasMoreResults) {
      return;
    }

    const scrollBottom = refElement.scrollHeight - refElement.offsetHeight - refElement.scrollTop;
    if (scrollBottom <= loadMoreDistanceFromBottom) {
      dispatch(loadMoreStudyGuides());
    }
  };

  return (
    <StudyGuidesBody
      ref={ref}
      onScroll={() => {
        const refElement = assertNotNull(ref.current, 'Scroll event before on non existing dom node');

        updateBackToTop(refElement);
        fetchMoreStudyGuides(refElement);
      }}
      data-testid='show-studyguides-body'
      data-analytics-region='SG popup'
    >
      <StudyGuidesCTA />
      <Filters />
      <StudyGuides />
      {showGoToTop && <GoToTopButton
        i18nAriaLabel='i18n:toolbar:studyguides:popup:button:back-to-top'
        onClick={goToTop}
        data-testid='back-to-top-studyguides'
      />}
    </StudyGuidesBody>
  );
};

export default ShowStudyGuides;
