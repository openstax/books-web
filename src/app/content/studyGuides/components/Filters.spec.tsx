import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import * as Cookies from 'js-cookie';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import { book as archiveBook } from '../../../../test/mocks/archiveLoader';
import { mockCmsBook } from '../../../../test/mocks/osWebLoader';
import { receiveLoggedOut, receiveUser } from '../../../auth/actions';
import Checkbox from '../../../components/Checkbox';
import { DropdownToggle } from '../../../components/Dropdown';
import * as Services from '../../../context/Services';
import MessageProvider from '../../../MessageProvider';
import { Store } from '../../../types';
import { assertWindow } from '../../../utils';
import FiltersList from '../../components/popUp/FiltersList';
import { formatBookData, stripIdVersion } from '../../utils';
import {
  printStudyGuides,
  receiveStudyGuidesTotalCounts,
  receiveSummaryStudyGuides,
  updateSummaryFilters,
} from '../actions';
import Filters from './Filters';
import { cookieUTG } from './UsingThisGuide/constants';
import UsingThisGuideBanner from './UsingThisGuide/UsingThisGuideBanner';
import UsingThisGuideButton from './UsingThisGuide/UsingThisGuideButton';

jest.mock('../../components/popUp/ChapterFilter', () => (props: any) => <div mock-chapter-filter {...props} />);

describe('Filters', () => {
  let store: Store;
  let services: ReturnType<typeof createTestServices>;
  let dispatch: jest.SpyInstance;
  const window = assertWindow();
  const book = formatBookData(archiveBook, mockCmsBook);

  beforeEach(() => {
    services = createTestServices();
    store = createTestStore();
    services = createTestServices();

    window.print = jest.fn();

    dispatch = jest.spyOn(store, 'dispatch');
  });

  it('matches snapshot with UTG banner open (opened initially)', () => {
    const pageId = stripIdVersion(book.tree.contents[0].id);
    store.dispatch(receiveStudyGuidesTotalCounts({
      [pageId]: {
        [HighlightColorEnum.Green]: 1,
        [HighlightColorEnum.Yellow]: 1,
        [HighlightColorEnum.Blue]: 1,
        [HighlightColorEnum.Pink]: 1,
        [HighlightColorEnum.Purple]: 1,
      },
    }));

    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <Filters />
        </MessageProvider>
      </Services.Provider>
    </Provider>);

    renderer.act(() => {
      const [chapterFilterToggle, colorFilterToggle] = component.root.findAllByType(DropdownToggle);
      chapterFilterToggle.props.onClick();
      colorFilterToggle.props.onClick();
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correct label keys for color dropdown', () => {
    const pageId = stripIdVersion(book.tree.contents[0].id);
    store.dispatch(receiveStudyGuidesTotalCounts({
      [pageId]: {
        [HighlightColorEnum.Green]: 1,
        [HighlightColorEnum.Yellow]: 1,
        [HighlightColorEnum.Blue]: 1,
        [HighlightColorEnum.Pink]: 1,
        [HighlightColorEnum.Purple]: 1,
      },
    }));

    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <Filters />
        </MessageProvider>
      </Services.Provider>
    </Provider>);

    const labelBlueKey = component.root.findByProps({ id: 'i18n:studyguides:popup:filters:blue' });
    const labelGreenKey = component.root.findByProps({ id: 'i18n:studyguides:popup:filters:green' });
    const labelPurpleKey = component.root.findByProps({ id: 'i18n:studyguides:popup:filters:purple' });
    const labelYellowKey = component.root.findByProps({ id: 'i18n:studyguides:popup:filters:yellow' });

    expect(labelBlueKey).toBeTruthy();
    expect(labelGreenKey).toBeTruthy();
    expect(labelPurpleKey).toBeTruthy();
    expect(labelYellowKey).toBeTruthy();
  });

  it('does renders ConnectedFilterList if user is logged in', () => {
    store.dispatch(receiveUser({} as any));
    const pageId = stripIdVersion(book.tree.contents[0].id);
    store.dispatch(receiveStudyGuidesTotalCounts({
      [pageId]: {
        [HighlightColorEnum.Green]: 1,
        [HighlightColorEnum.Yellow]: 1,
      },
    }));

    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <Filters />
        </MessageProvider>
      </Services.Provider>
    </Provider>);

    expect(() => component.root.findByType(FiltersList)).not.toThrow();
  });

  it('does not render ConnectedFilterList if user is logged out', () => {
    store.dispatch(receiveLoggedOut());
    const pageId = stripIdVersion(book.tree.contents[0].id);
    store.dispatch(receiveStudyGuidesTotalCounts({
      [pageId]: {
        [HighlightColorEnum.Green]: 1,
        [HighlightColorEnum.Yellow]: 1,
      },
    }));

    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <Filters />
        </MessageProvider>
      </Services.Provider>
    </Provider>);

    expect(() => component.root.findByType(FiltersList)).toThrow();
  });

  it('dispatches updateSummaryFilters action', () => {
    const pageId = stripIdVersion(book.tree.contents[0].id);
    store.dispatch(receiveStudyGuidesTotalCounts({
      [pageId]: {
        [HighlightColorEnum.Green]: 1,
        [HighlightColorEnum.Yellow]: 1,
      },
    }));

    const component = renderer.create(<Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <Filters />
        </MessageProvider>
      </Services.Provider>
    </Provider>);

    renderer.act(() => {
      const [, colorFilterToggle] = component.root.findAllByType(DropdownToggle);
      colorFilterToggle.props.onClick();
    });

    const [yellowCheckbox] = component.root.findAllByType(Checkbox);

    renderer.act(() => {
      yellowCheckbox.props.onChange();
    });

    expect(dispatch).toHaveBeenCalledWith(updateSummaryFilters({
      colors: { new: [], remove: [HighlightColorEnum.Yellow] },
    }));

    renderer.act(() => {
      yellowCheckbox.props.onChange();
    });

    expect(dispatch).toHaveBeenCalledWith(updateSummaryFilters({
      colors: { new: [HighlightColorEnum.Yellow], remove: [] },
    }));
  });

  describe('PrintButton', () => {
    it('triggers print immediately if nothing left to load', () => {
      const component = renderer.create(<Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider>
            <Filters />
          </MessageProvider>
        </Services.Provider>
      </Provider>);

      const printButton = component.root.findByProps({'data-testid': 'print'});

      renderer.act(() => {
        printButton.props.onClick();
      });

      expect(dispatch).not.toHaveBeenCalledWith(printStudyGuides());
      expect(window.print).toHaveBeenCalled();
    });

    it('loads remaining highlights before printing', () => {
      store.dispatch(receiveSummaryStudyGuides({}, {pagination: {} as any}));

      const component = renderer.create(<Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider>
            <Filters />
          </MessageProvider>
        </Services.Provider>
      </Provider>);

      const printButton = component.root.findByProps({'data-testid': 'print'});

      renderer.act(() => {
        printButton.props.onClick();
      });

      expect(dispatch).toHaveBeenCalledWith(printStudyGuides());
    });
  });

  describe('Using This Guide Button', () => {
    it('renders button and banner when button is clicked and closes correctly', () => {
      // If cookie is set then banner will be closed initially
      Cookies.set(cookieUTG, 'true');

      const component = renderer.create(<Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider>
            <Filters />
          </MessageProvider>
        </Services.Provider>
      </Provider>);

      const uTGbutton = component.root.findByType(UsingThisGuideButton);

      renderer.act(() => {
        uTGbutton.props.onClick();
      });

      expect(component.root.findByType(UsingThisGuideBanner).props.show).toBe(true);

      const uTGcloseButton = component.root.findByProps({ 'data-testid': 'close-utg' });

      renderer.act(() => {
        uTGcloseButton.props.onClick();
      });

      expect(component.root.findByType(UsingThisGuideBanner).props.show).toBe(false);
    });

    it('does not send ga event if it was opened initialy but set cookie', () => {
      // If cookie is not set then banner will be opened initially
      Cookies.remove(cookieUTG);

      const spyTrack = jest.spyOn(services.analytics.openUTG, 'track');

      const component = renderer.create(<Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider>
            <Filters />
          </MessageProvider>
        </Services.Provider>
      </Provider>);

      // wait for hooks
      // tslint:disable-next-line: no-empty
      renderer.act(() => {});

      const banner = component.root.findByType(UsingThisGuideBanner);
      expect(banner.props.show).toEqual(true);
      expect(Cookies.get(cookieUTG)).toBe('true');
      expect(spyTrack).not.toHaveBeenCalled();
    });

    it('send ga event when opened', () => {
      const spyTrack = jest.spyOn(services.analytics.openUTG, 'track');
      // If cookie is set then banner will be closed initially
      Cookies.set(cookieUTG, 'true');

      const component = renderer.create(<Provider store={store}>
        <Services.Provider value={services}>
          <MessageProvider>
            <Filters />
          </MessageProvider>
        </Services.Provider>
      </Provider>);

      const banner = component.root.findByType(UsingThisGuideBanner);
      expect(banner.props.show).toEqual(false);
      expect(Cookies.get(cookieUTG)).toEqual('true');
      expect(spyTrack).not.toHaveBeenCalled();

      const toggleButton = component.root.findByType(UsingThisGuideButton);

      // open banner
      renderer.act(() => {
        toggleButton.props.onClick();
      });

      expect(banner.props.show).toEqual(true);
      expect(spyTrack).toHaveBeenCalledTimes(1);

      // close banner
      renderer.act(() => {
        toggleButton.props.onClick();
      });

      expect(banner.props.show).toEqual(false);
      expect(spyTrack).toHaveBeenCalledTimes(1); // do not send ga event on close

      // open banner again
      renderer.act(() => {
        toggleButton.props.onClick();
      });

      expect(banner.props.show).toEqual(true);
      expect(spyTrack).toHaveBeenCalledTimes(2); // send ga event every time when opening
    });
  });
});
