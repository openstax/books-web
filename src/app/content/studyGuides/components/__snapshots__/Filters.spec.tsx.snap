// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters matches snapshot 1`] = `
<div
  className="Filters-sc-1fn81fx-2 cJUIZf"
>
  <div
    className="Dropdown-rmc6yw-5 fBPFfC Dropdown__TabHiddenDropDown-rmc6yw-1 kbMrSv"
  >
    <button
      aria-label="Filter study guides by Chapter"
      className="Dropdown__DropdownToggle-rmc6yw-0 gpiKKF Filters__Toggle-sc-1fn81fx-1 dpyUXV Button__PlainButton-ayg7nk-2 chgswm"
      onClick={[Function]}
    >
      <div
        tabIndex={-1}
      >
        Chapter
        <svg
          aria-hidden="true"
          className="Filters__DownIcon-sc-1fn81fx-0 hoQnJB StyledIconBase-sc-bdy9j4 cCvWXa"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 320 512"
        >
          <path
            d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
            fill="currentColor"
          />
        </svg>
      </div>
    </button>
    <div
      disabled={false}
      locationFilters={Map {}}
      locationFiltersWithContent={Set {}}
      mock-chapter-filter={true}
      selectedLocationFilters={Set {}}
      setFilters={[Function]}
    />
  </div>
  <div
    className="ColorKey__ColorKeyWrapper-afl82f-6 eURUEn"
  >
    <button
      aria-label="Color key"
      className="Button__PlainButton-ayg7nk-2 ColorKey__ColorKeyButtonWrapper-afl82f-0 fkFtKM"
      onClick={[Function]}
    >
      <img
        className="ColorKey__ColorKeyIcon-afl82f-2 kbHiQJ"
        src="colorKeyIcon.png"
      />
      <span
        className="ColorKey__ColorKeyText-afl82f-3 cLeOIZ"
      >
        Color key
      </span>
    </button>
  </div>
  <div
    className="UsingThisGuideButton__UsingThisGuideWrapper-bu0xmw-3 eHjhkf"
  >
    <button
      aria-label="Using This Guide"
      className="Button__PlainButton-ayg7nk-2 UsingThisGuideButton__UsingThisGuideButtonWrapper-bu0xmw-1 fTisnp"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="UsingThisGuideButton__QuestionIcon-bu0xmw-0 hrjHGl StyledIconBase-sc-bdy9j4 cCvWXa"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 512 512"
      >
        <path
          d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
          fill="currentColor"
        />
      </svg>
      <span
        className="UsingThisGuideButton__UsingThisGuideText-bu0xmw-2 jFOwWV"
      >
        Using This Guide
      </span>
    </button>
  </div>
  <button
    aria-label="print"
    className="styled__PlainButton-sc-1119auh-0 PrintButton-sc-1bdptk0-0 byafPq styled__PrintOptWrapper-sc-1119auh-1 kbtaBS"
    data-testid="print"
    disabled={false}
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="styled__PrintIcon-sc-1119auh-3 iKTgDK StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 512 512"
    >
      <path
        d="M448 192V77.25c0-8.49-3.37-16.62-9.37-22.63L393.37 9.37c-6-6-14.14-9.37-22.63-9.37H96C78.33 0 64 14.33 64 32v160c-35.35 0-64 28.65-64 64v112c0 8.84 7.16 16 16 16h48v96c0 17.67 14.33 32 32 32h320c17.67 0 32-14.33 32-32v-96h48c8.84 0 16-7.16 16-16V256c0-35.35-28.65-64-64-64zm-64 256H128v-96h256v96zm0-224H128V64h192v48c0 8.84 7.16 16 16 16h48v96zm48 72c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z"
        fill="currentColor"
      />
    </svg>
    <span
      className="styled__PrintOptions-sc-1119auh-2 bRCGrc"
    >
      Print
    </span>
  </button>
  <ul
    className="FiltersList-sc-1sdnk9e-3 fQoHdD"
  />
</div>
`;
